src/cockpit/beiboot.py

```diff
@@ -256,7 +263,11 @@ async def connect_from_bastion_host(self) -> None:
         if not ssh_askpass.exists():
             logger.error("Could not find cockpit-askpass helper at %r", askpass)
 
-        cmd, env = via_ssh(cmd, self.destination, ssh_askpass, *username_opts)
+        if known_hosts is not None:
```

Comment:
This feels slightly wrong.  I think we always want to set this.  Even if the user didn't specify it to us in the first place, we will want to have the tempfile there to report results back to the user.

I'm guessing the `is None` path is never actually hit since whenever you run this code the login page is unconditionally sending us `'\0'` in our auth request?  This could use some clarification in any case...

btw: it would be nice to move this stuff to ferny eventually.

Reply:
I started with always setting it, but that's not compatible with the old login page which did not set it (that caused several test failures, and was one of the corner cases which took me a long time to get right). But moreover, we don't set this in Client mode, and we really really *don't* want to set a custom temp knownhosts file then -- it should use the user's real `~/.ssh/known_hosts`.

Reply:
I added a big comment explaining it.

Reply:
This is inside of `connect_from_bastion_host()` — it's not run in the Cockpit Client case.

Also, I don't think we have to worry about compatibility with the old login page.  `cockpit-ws`, `pkg/static/` and `cockpit-beiboot` will always be shipped together.

Reply:
It's an interesting point about the tmpdir scope, though: we shouldn't create the tempdir when we're in the flatpak case.  So don't move that to `__init__()`....

Reply:
> it's not run in the Cockpit Client case.

Good point -- however, it still feels wrong to me to hide the user's ~/.ssh/known_hosts if the caller didn't explicitly ask for it. With the recent test adjustments to TestClient etc. it'll probably work, but it does feel unclean to me. If you prefer, I can change it though.

> we shouldn't create the tempdir when we're in the flatpak case.

Doesn't hurt much, does it? Current push has it in __init__, but I can move it back of course (no strong opinion)

Reply:
TestClient and testBeibootWithBridge both pass with this:

```diff
--- src/cockpit/beiboot.py
+++ src/cockpit/beiboot.py
@@ -279,12 +279,7 @@ class SshPeer(Peer):
 
         if response.startswith('Basic '):
             decoded = base64.b64decode(response[6:]).decode()
-            user_password, sep, data = decoded.partition('\0')
-            # be careful: not specifying known-hosts data at all is different from an empty set of keys
-            # the former means "use user's default ~/.ssh/config" (for Client), the latter uses a temporary file
-            # (for bastion host mode of the login page)
-            if sep:
-                known_hosts = data
+            user_password, _, known_hosts = decoded.partition('\0')
             user, _, basic_password = user_password.partition(':')
             if user:  # this can be empty, i.e. auth is just ":"
                 logger.debug("got username %s and password from Basic auth", user)
```

So it would *work*, but I still don't prefer it -- I'd rather be a bit more careful here.

Reply:
Here's my take: we have two (and a half) scenarios that this code will run in:
 - flatpak for Cockpit Client (will not run this function)
 - cockpit-ssh for bastion host: in this case we're actually running ssh as the cockpit-wsinstance user on behalf of someone logging in via the web and we never ever want it to have access to `~/.ssh`, which hopefully doesn't even exist.  We absolutely always want to create an isolated environment for this user.  (mental note: maybe we should drop privs in this case...)
   - (the half case) same as above, but inside of a container

Reply:
OK -- I applied the above patch.

Reply:
But what's the scenario where we accept that we didn't receive a `Basic` auth, then?  You still allow that `known_hosts` may be None, which can now only happen if we skip that entire block...

Recall: this scenario is only for bastion host.  I think we should *always* have a Basic auth in that case, no?

I could have probably done some more simplification when I split these two paths... but this commit ended up coming 'after' in the series...

Reply:
As I said, it will work (tests are green). I don't know if there's third-party scenarios which speak that protocol (foreman? etc.). But we'll know :grin: And these should also not use ~/.ssh/